abci/check_tx.go

```diff
@@ -0,0 +1,186 @@
+		// bid transactions.
+		txDecoder sdk.TxDecoder
+
+		// mempool is utilized to retrieve the bid info of a transaction and to
+		// insert a transaction into the application-side mempool.
+		mempool CheckTxMempool
+
+		// anteHandler is utilized to verify the bid transaction against the latest
+		// committed state.
+		anteHandler sdk.AnteHandler
+
+		// chainID is the chain ID of the blockchain.
+		chainID string
+	}
+
+	// CheckTx is baseapp's CheckTx method that checks the validity of a
+	// transaction.
+	CheckTx func(cometabci.RequestCheckTx) cometabci.ResponseCheckTx
+
+	// CheckTxMempool is the interface that defines all of the dependencies that
+	// are required to interact with the application-side mempool.
+	CheckTxMempool interface {
+		// GetAuctionBidInfo is utilized to retrieve the bid info of a transaction.
+		GetAuctionBidInfo(tx sdk.Tx) (*mempool.AuctionBidInfo, error)
+
+		// Insert is utilized to insert a transaction into the application-side mempool.
+		Insert(ctx context.Context, tx sdk.Tx) error
+
+		// WrapBundleTransaction is utilized to wrap a transaction included in a bid transaction
+		// into an sdk.Tx.
+		WrapBundleTransaction(tx []byte) (sdk.Tx, error)
+	}
+
+	// BaseApp is an interface that allows us to call baseapp's CheckTx method
+	// as well as retrieve the latest committed state.
+	BaseApp interface {
+		// CommitMultiStore is utilized to retrieve the latest committed state.
+		CommitMultiStore() sdk.CommitMultiStore
+
+		// CheckTx is baseapp's CheckTx method that checks the validity of a
+		// transaction.
+		CheckTx(cometabci.RequestCheckTx) cometabci.ResponseCheckTx
+
+		// Logger is utilized to log errors.
+		Logger() log.Logger
+
+		// LastBlockHeight is utilized to retrieve the latest block height.
+		LastBlockHeight() int64
+	}
+)
+
+// NewCheckTxHandler is a constructor for CheckTxHandler.
+func NewCheckTxHandler(baseApp BaseApp, txDecoder sdk.TxDecoder, mempool CheckTxMempool, anteHandler sdk.AnteHandler, chainID string) CheckTxHandler {
+	return CheckTxHandler{
+		baseApp:     baseApp,
+		txDecoder:   txDecoder,
+		mempool:     mempool,
+		anteHandler: anteHandler,
+		chainID:     chainID,
+	}
+}
+
+// CheckTxHandler is a wrapper around baseapp's CheckTx method that allows us to
+// verify bid transactions against the latest committed state. All other transactions
+// are executed normally. We must verify each bid tx and all of its bundled transactions
+// before we can insert it into the mempool against the latest commit state because
+// otherwise the auction can be griefed. No state changes are applied to the state
+// during this process.
+func (handler *CheckTxHandler) CheckTx() CheckTx {
+	return func(req cometabci.RequestCheckTx) (resp cometabci.ResponseCheckTx) {
+		defer func() {
+			if err := recover(); err != nil {
+				resp = sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("panic in check tx handler: %s", err), 0, 0, nil, false)
+			}
+		}()
```

Comment:
I know that base app's `CheckTx` sets up a bunch of different middleware in the case when the transaction panics at any point. Do you think this is sufficient enough or should we try to mirror the same defer logic that is on `runTx`?